New tutorial section on how to use resources (#14566)
dagster-io/dagster: docs/content/tutorial/connecting-to-external-services.mdx

---

> title: "Tutorial, part seven: Connecting to external services | Dagster Docs"

Reviewer: ```suggestion
title: "Tutorial, part eight: Connecting to external services | Dagster Docs"
```

description: Learn how to connect to external services in Dagster with resources

> ---
> 

Reviewer: ```suggestion
---

# Tutorial, part 8: Connecting to external services

```

⛏️ Missing the page title here

> Throughout the tutorial, you have built a data pipeline that connects to different external services, such as Hacker News. By building this asset graph, Dagster acts as a control plane of multiple different external services. However, connecting to these services and managing their use can be difficult. Resources are the recommended way to manage these connections in Dagster.

Reviewer: ```suggestion
Throughout the tutorial, you built a data pipeline that connects to different external services, such as Hacker News. By building this asset graph, Dagster acts as a control plane of multiple different external services. However, connecting to these services and managing their use can be difficult. Resources are the recommended way to manage these connections in Dagster.
```

💭 Suggesting we remove the `have` from `you have built` to make this more active and concise

> Resources are Dagster objects that connect to external services, such as Slack, Snowflake, or S3. In this section, you’ll learn more about resources, including how to:

Reviewer: ```suggestion
Resources are Dagster objects that connect to external services, such as Slack, Snowflake, or Amazon S3. In this section, you’ll learn more about resources, including how to:
```

⛏️ The first time a product like this is mentioned, the name of the company/provider/etc should be included. After, it can just be the product - ex: `S3`

- Define a resource and use it in an asset
- Configure a resource
- Customize a resource based on the environment where Dagster is running

> ## Why use resources?

Reviewer: ```suggestion
---

## Why use resources?
```

In the `topstory_ids` and `topstories` assets, you used the `requests` library to communicate with the Hacker News API. This code gets harder to maintain if:

- Connections need to be shared across multiple assets

> - You need to authenticate with the external service

Reviewer 2: ```suggestion
- You need to authenticate with the external service with secrets stored safely
```

- Configurations need to be consistent across its usage

> - If any of the above changes based on environment (e.g. development vs. production)

Reviewer 2: may worth calling out something like vary tokens in dev vs prod to avoid rate limit. and may also worth saying something about tests as resources play an important role in unit tests.

> Aside from the complexity of configuring them, building data pipeline creates many connections to external services. It becomes difficult to understand what assets use which connections or how often a connection is used.

Reviewer: ```suggestion
Aside from the complexity of configuring them, building a data pipeline creates many connections to external services. It becomes difficult to understand what assets use which connections or how often a connection is used.
```

You should use resources to manage communicating with external services because they:

- Enable flexibility in configuration
- Standardize how connections are used across assets
- Create a single source to monitor connections in the Dagster UI

---

> # Step 1: Setting up a resource

Reviewer: ```suggestion
## Step 1: Setting up a resource
```

> Suppose your Hacker News pipeline in Dagster has picked up momentum, and your stakeholders want to learn more about who is using it. You are tasked with analyzing how many people sign up for Hacker News.

Reviewer 2: worth showing what it looks like without using resource, so that it's more obvious about why resources help in this case with before (request api direct invoked in asset) vs after (using resource + no change in asset body when switching resource/config)

> In the scaffolded Dagster project you made in Part 2 of the tutorial, you'll have seen a directory called `resources` with some files in it. The most important file is the `resources.py` file that exposes a resource called `DataGeneratorResource`. This resource generates simulated data about Hacker News signups. You'll use this resource to get the data needed to produce an asset for analysis.

Reviewer: ```suggestion
The scaffolded Dagster project you made in [Part 2 of the tutorial](/tutorial/setup), contains a directory called `resources` with some files in it. The most important file is the `resources.py` file that exposes a resource called `DataGeneratorResource`. This resource generates simulated data about Hacker News signups. You'll use this resource to get the data needed to produce an asset for analysis.
```

💭 Suggesting a few changes for structure and conciseness

<Note>
  The signup data from Hacker News is fake and generated by the library. This is
  simulated data and should not be used for real use cases.
</Note>

> In your `__init__.py`, import the class, create an instance of it, and add it to the `resources` argument for your code location's `Definitions` object under the key `hackernews_api`. The key used to define a resource in the `Definitions` object is the key you'll use to reference the resource later in your code. In this case, we’ll call it `hackernews_api`.

Reviewer: Should `Definitions` use the PyObject component?

Verify that your code looks similar to the code below:

> ```python

Reviewer 2: do we plan to check in those code snippets somewhere so we can keep stuff tested and up-to-date?
maybe alongside with https://github.com/dagster-io/dagster/pull/14560 (and maybe name it as `tutorial_template`?), we have a checked-in complete version called `tutorial_completed"?

Author: 🖖🏽 Yeup! The plan is to check in and test the snippets once we can agree on content, in case anything gets checked

Also great idea with putting a `tutorial_completed` template!

from .resources import DataGeneratorResource

# ...

datagen = DataGeneratorResource() # Make the resource

defs = Definitions(
    assets=[*hackernews_assets],
    schedules=[hackernews_schedule],
    resources={
        "io_manager": io_manager,
        "database_io_manager": duckdb_io_manager,
        "hackernews_api": datagen # Add the newly-made resource here
    }
)
```

To confirm that this worked:

1. Go into the Dagster UI
2. Reload your definitions
3. Navigate to the **Overview** page
4. Go to the **Resources** tab
5. Find the `hackernews_api` resource in the tab and click on it
6. Observe the details of the resource

![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/d1aaa68c-cb55-4805-a797-66dcdb672d5f/Untitled.png)

<Image

> alt="the resources overview page in the Dagster UI"

Reviewer: ```suggestion
alt="The Resources overview page in the Dagster UI"
```

⛏️ Page titles should be capitalized

src="/images/tutorial/resources-overview.png"
width={1152}
height={610}
/>

---

# Step 2: Using the resource

On this page, you might notice that the **Uses** value for the `hackernews_api` resource is 0. This is because you've defined the resource but have yet to use it.

In your `assets.py`, make a new asset called `signups`. In this asset, you'll use your newly made resource to get the simulated data and return back the signups.

Copy and paste the code below into the appropriate sections of your `asset.py`.

```python
from ..resources import DataGeneratorResource

# ...

@asset
def signups(hackernews_api: DataGeneratorResource):
    signups = pd.DataFrame(hackernews_api.get_signups())

    return Output(

> 				value=signups,

Reviewer 2: indentation is off here

				metadata={
				  	"Record Count": len(signups),
					  "Preview": MetadataValue.md(signups.head().to_markdown()),
		        "Earliest Signup": signups["registered_at"].min(),
		        "Latest Signup": signups["registered_at"].max(),
		}
	)
```

Verify this new asset works by reloading your definitions and materializing the `signups` asset. If you navigate to the `hackernews_api`'s resource page from earlier, you’ll notice its **Uses** value now says `2`: one for the asset using it and another for the job using this asset. Click on the number and ensure that your `signups` asset uses the `hackernews_api` resource.

<Image
alt="indicating how many times a resource is used in the Dagster UI"
src="/images/tutorial/resources-uses.png"
width={1152}
height={610}
/>

---

> # Step 3: Configuring the resource

Reviewer: ```suggestion
## Step 3: Configuring the resource
```

Resources can take arguments to configure them. However, this resource lets you define how many days of data should be fetched. Let’s say you want to get a year’s worth of data. Update your instantiation of the `DataGeneratorResource` to match the change below:

```python
defs = Definitions(
	#...
	resources={
		"hackernews_api": DataGeneratorResource(num_days=365)
	}
)
```

Confirm this worked by materializing the `signups` asset again. The `Earliest Signup` date should be a year ago from the `Latest Signup` date.

<Image
alt="the Dagster asset graph with metadata about when the earliest signup was"
src="/images/tutorial/resources-materialization.png"
width={1152}
height={610}
/>

## Configuring based on the environment

Imagine that you've deployed this new asset to your production Dagster instance. As you're tinkering around in development, you start thinking that you don't need a year's worth of data during local development. By fetching all signups in the past year, materializing the asset is slow and takes up a lot of storage on your computer. However, your stakeholders still need a whole year's worth of data.

> By using environment variables, you can configure how the `DataGeneratorResource` behaves based on the environment where Dagster is running. In development, you can fetch only the past 30 days of data for a rapid feedback loop. For production, you can get the full year of data needed.

Reviewer: ```suggestion
By using [environment variables](/guides/dagster/using-environment-variables-and-secrets), you can configure how the `DataGeneratorResource` behaves based on the environment where Dagster is running. In development, you can fetch only the past 30 days of data for a rapid feedback loop. For production, you can get the full year of data needed.
```

❓ Thoughts on adding a link to the env vars guide here?

To do this, create a new file called `.env` in the directory you run `dagster dev` from, which is usually the root-level directory of your Dagster project. This is where you'll put your environment variables. Copy and paste the following line into the `.env` file:

```bash
HACKERNEWS_NUM_DAYS_WINDOW=30
```

Afterward, you'll use Dagster's `EnvVar` class to access the environment variable. Update your `__init__.py` with the changes featured below:

```python
from dagster import EnvVar

# ...

defs = Definitions(
	#...
	resources={
		"hackernews_api": DataGeneratorResource(
			num_days=EnvVar.int("HACKERNEWS_NUM_DAYS_WINDOW"),
		)
	}
)
```

Restart your Dagster instance, materialize your signups asset, and observe how the `Earliest Signup` metadata value is 30 days ago from today.

> # Next Steps

Reviewer: ```suggestion
---

# Next steps
```

Congratulations! 🎉 You've written your first pipeline in Dagster!

> In this section, you used your first resource to manage your connection to an external source. Many resources are available in the Dagster ecosystem, such as resources for executing queries against Snowflake, sending messages to Slack, or executing a Jupyter notebook. You can also create resources to manage connections to your own services or software that isn't natively integrated with Dagster.

Reviewer 2: should we link to relevant docs about resources here?
